rollo-printer: init at 1.8.4

shymega opened the pull request:
This PR adds the Rollo label printer CUPS driver - which is also used by some Munbyn label printers. It currently only works over USB.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

shymega committed:
rollo-printer: init at 1.8.4

Yarny0 commented:
Hi @shymega , I was about to add a review, then I discovered that the [`homepage`](https://www.rollo.com/driver-linux/) also provides a "Souce tarball".  I think nixpkgs tries to build from source whenever possible.  Is it possible to compile the driver from that source tarball?  That would also increase the probability that it can be used on other architectures.

shymega commented:
> Hi @shymega , I was about to add a review, then I discovered that the [`homepage`](https://www.rollo.com/driver-linux/) also provides a "Souce tarball". I think nixpkgs tries to build from source whenever possible. Is it possible to compile the driver from that source tarball? That would also increase the probability that it can be used on other architectures.

Building from source didn't work. See my mention in the PR description. The issue was that despite setting the right variables, the source tarball still attempted to install into the Nix store, under the CUPS directory. I didn't want to make this more complex by patching, so went with the binary.

shymega commented:
Hi @Yarny0

Good shout, that did the trick. I tested the output, and it seems to work. I don't run NixOS on my print server for my label printer, but I will eventually, and the PPD and compiled binary look good to me.

Yarny0 commented on pkgs/by-name/ro/rollo-cups-driver/package.nix:
```diff
@@ -0,0 +1,43 @@
+  stdenv,
+}:
+let
+  pname = "rollo-cups-driver";
```
As I researched, the package is known to Repology under the name [`rollo-printer`](https://repology.org/projects/?search=rollo) only (and is only packaged by [AUR](https://repology.org/project/rollo-printer/versions) at the time of this writing).  Maybe it is a good idea to call our package also `rollo-printer` (here and the name of the `by-name` directory) so Repology can make the connection more easily.

Yarny0 commented on pkgs/by-name/ro/rollo-cups-driver/package.nix:
```diff
@@ -0,0 +1,43 @@
+    cups
+  ];
+
+  makeFlags = [
```
Can you add a comment that explains why this is necessary?

Yarny0 replied:
Oh sorry for confusion: My comment about adding a comment referred to `makeFlags`.  (I marked the `makeFlags` line, but the way Github displays it here is a bit ambiguous.  It is displayed clearer in the "Files changed" tab.)

I agree that the comment about `buildInputs` is not needed.

Yarny0 reviewed (changes requested):
#### Review per https://github.com/NixOS/nixpkgs/tree/00c21e4c93d963c50d4c0c89bfa84ed6e0694df2/pkgs#new-packages

Everything below observed/built/tested after cherry-picking the commit onto [current `nixos-unstable`](https://github.com/NixOS/nixpkgs/tree/00c21e4c93d963c50d4c0c89bfa84ed6e0694df2).

##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on `x86_64-linux`
- [x] executables tested on `x86_64-linux`
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] `meta.mainProgram` is set, if applicable.  (*not applicable here*)
- [x] build time only dependencies are declared in `nativeBuildInputs`  (*none*)
- [x] source is fetched from an official or trusted location
- [ ] source is fetched using the appropriate function  (*see comments*)
- [ ] the motives for any special packaging choices are documented  (*see comments*)
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.  (*none*)
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed  (*none*)
- [x] patches that are remotely available are fetched rather than vendored  (*none*)

More:

- [x] CUPS produces printer data without reporting any error, after installing the ppd file with `hardware.ensurePrinters` (tested by printing into a file as I have no Rollo printer)

##### Possible improvements

*   Very minor nitpick: [`init` in the commit title should not be capitalized](https://github.com/NixOS/nixpkgs/tree/00c21e4c93d963c50d4c0c89bfa84ed6e0694df2/pkgs#commit-conventions).

Yarny0 commented on pkgs/by-name/ro/rollo-printer/package.nix:
```diff
@@ -0,0 +1,32 @@
+  fetchurl,
+  lib,
+  stdenv,
+}: stdenv.mkDerivation (finalAttrs: {
```
`nixfmt` wants a line break here (before `stdenv.mkDerivation`).  That why CI checks currently fail.

Yarny0 reviewed (approved):
This looks very convincing now.  I redid my tests and the printer driver still produces output data without any errors.  In my opinion this is ready for merging.

SuperSandro2000 commented on pkgs/by-name/ro/rollo-printer/package.nix:
```diff
@@ -0,0 +1,37 @@
+    hash = "sha256-v61/5HY25cvhVbHF+dXOOGrDfZZzvvicJEy7MKTAG10=";
+  };
+
+  # The configure script requires CUPS as a build input during the buildPhase.
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/ro/rollo-printer/package.nix:
```diff
@@ -0,0 +1,37 @@
+  # These are necessary to set, otherwise the Makefile will try to write to the
+  # Nix store directory of the `cups` build input. Setting these variables
+  # contains the Makefile to write within the `$out` directory, as would be
+  # expected.
```
```suggestion
```

SuperSandro2000 reviewed (commented):
I don't think we need those comments as both are expected and the lines below them do not raise any questions out of the usual.

shymega commented on pkgs/by-name/ro/rollo-printer/package.nix:
```diff
@@ -0,0 +1,37 @@
+    homepage = "https://www.rollo.com/driver-linux/";
+    license = with lib.licenses; gpl3Plus;
+    maintainers = with lib.maintainers; [ shymega ];
+    platforms = lib.platforms.all;
```
_This_ is a really good idea. I've applied it. Should've thought of this.

shymega commented:
Ah, so the driver only appears to compile under x86_64-linux. This seems in contradiction to the driver's web page, and especially as I can run the binary package on a Raspberry Pi 3.

I wonder if we need to do something specific in the `makeFlags` to fix this under non-x86 architectures?

Yarny0 commented:
[@shymega wrote](https://github.com/NixOS/nixpkgs/pull/464258#issuecomment-3948029420):
> I wonder if its viable to rely on Rollo's CDN for this derivation long-term though?

I'd say yes: [AUR is also using it](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=rollo-printer&id=c61d2594cc6cf754a3c6a96cb4aa30a12d9a864b#n13) and one successful download will be enough for our binary cache -- unless there is an update, of course.

SuperSandro2000 commented:
```
Running phase: unpackPhase
@nix {"action":"setPhase","phase":"unpackPhase"}
unpacking source archive /nix/store/d5mxffrar52v2mi0nj0zcgskz89x1xmk-rollo-cups-driver-1.8.4.tar.gz
source root is rollo-cups-driver-1.8.4
setting SOURCE_DATE_EPOCH to timestamp 1718644355 of file "rollo-cups-driver-1.8.4/rollo-x1038.ppd"
Running phase: patchPhase
@nix {"action":"setPhase","phase":"patchPhase"}
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix {"action":"setPhase","phase":"updateAutotoolsGnuConfigScriptsPhase"}
Updating Autotools / GNU config script to a newer upstream version: ./config.sub
Updating Autotools / GNU config script to a newer upstream version: ./config.guess
Running phase: configurePhase
@nix {"action":"setPhase","phase":"configurePhase"}
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/3zrx6av2d1141igkcn8477cvbfqpcmcf-bash-5.3p9/bin/sh"
configure flags: --prefix=/nix/store/kr21imc5ggxqnl1i7fz333gx3blvkvk6-rollo-printer-1.8.4
checking build system type... aarch64-apple-darwin24.6.0
checking host system type... aarch64-apple-darwin24.6.0
checking for gcc... clang
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether clang accepts -g... yes
checking for clang option to enable C11 features... none needed
checking for rm... /nix/store/chdj76m79c3lbwdn6bflwlqa2n2igx02-coreutils-9.10/bin/rm
checking for install-sh script... using /nix/var/nix/builds/nix-51712-1521736310/rollo-cups-driver-1.8.4/install-sh
checking for pkg-config... no
checking for cups-config... /nix/store/ckfkrrhls4rhpyi1026gz30mvr8jzzry-cups-2.4.16-dev/bin/cups-config
checking for CUPS library v2.2 or higher... ./configure: line 3621: --exists: command not found
yes, v2.4.16
checking for CUPS datadir... /nix/store/aixwwqkzsmkfs61nvbmyhjbvsz2jffvf-cups-2.4.16/share/cups
checking for CUPS serverbin... /nix/store/aixwwqkzsmkfs61nvbmyhjbvsz2jffvf-cups-2.4.16/libexec/cups
checking whether compiler supports -Wno-char-subscripts... yes
checking whether compiler supports -Wno-deprecated-declarations... yes
checking whether compiler supports -Wno-format-truncation... yes
checking whether compiler supports -Wno-format-y2k... yes
checking whether compiler supports -Wno-switch... yes
checking whether compiler supports -Wno-unused-result... yes
checking whether compiler supports -fPIE... yes
checking for OS-specific compiler options... -mmacosx-version-min=10.10 -arch x86_64 -arch arm64
configure: creating ./config.status
config.status: creating Makefile
Running phase: buildPhase
@nix {"action":"setPhase","phase":"buildPhase"}
build flags: SHELL=/nix/store/3zrx6av2d1141igkcn8477cvbfqpcmcf-bash-5.3p9/bin/bash CUPS_DATADIR=/nix/store/kr21imc5ggxqnl1i7fz333gx3blvkvk6-rollo-printer-1.8.4/share/cups CUPS_SERVERBIN=/nix/store/kr21imc5ggxqnl1i7fz333gx3blvkvk6-rollo-printer-1.8.4/lib/cups
Compiling rastertorollo.c...
Compiling rollo-dither.c...
Linking rastertorollo...
ld: warning: ignoring file /nix/store/vl936sd1vj8s008akm3gibkgyqk30qhz-clang-wrapper-21.1.8/resource-root/lib/darwin/libclang_rt.osx.a, building for macOS-x86_64 but attempting to link with file built for Mac Catalyst-arm64
ld: warning: ignoring file /nix/store/f19dskpf1ah7v5ysmm5y11x1w78yjfjw-cups-2.4.16-lib/lib/libcups.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
Undefined symbols for architecture x86_64:
  "_cupsFreeOptions", referenced from:
      _main in rastertorollo.o
  "_cupsMarkOptions", referenced from:
      _main in rastertorollo.o
  "_cupsParseOptions", referenced from:
      _main in rastertorollo.o
  "_cupsRasterClose", referenced from:
      _main in rastertorollo.o
  "_cupsRasterOpen", referenced from:
      _main in rastertorollo.o
  "_cupsRasterReadHeader2", referenced from:
      _main in rastertorollo.o
  "_cupsRasterReadPixels", referenced from:
      _main in rastertorollo.o
  "_ppdClose", referenced from:
      _main in rastertorollo.o
  "_ppdErrorString", referenced from:
      _main in rastertorollo.o
  "_ppdFindMarkedChoice", referenced from:
      _start_page in rastertorollo.o
      _end_page in rastertorollo.o
  "_ppdLastError", referenced from:
      _main in rastertorollo.o
  "_ppdMarkDefaults", referenced from:
      _main in rastertorollo.o
  "_ppdOpenFile", referenced from:
      _main in rastertorollo.o
ld: symbol(s) not found for architecture x86_64
clang: [0;1;31merror: [0m[1mlinker command failed with exit code 1 (use -v to see invocation)[0m
make: *** [Makefile:94: rastertorollo] Error 1
```

SuperSandro2000 commented on pkgs/by-name/ro/rollo-printer/package.nix:
```diff
@@ -0,0 +1,36 @@
+    homepage = "https://www.rollo.com/driver-linux/";
+    license = with lib.licenses; gpl3Plus;
+    maintainers = with lib.maintainers; [ shymega ];
+    platforms = cups.meta.platforms;
```
```suggestion
    platforms = cups.meta.platforms;
    broken = stdenv.hostPlatform.isDarwin;
```

shymega replied:
Good shout!

Yarny0 reviewed (approved):
#### Review per https://github.com/NixOS/nixpkgs/tree/6c9a78c09ff4d6c21d0319114873508a6ec01655/pkgs#new-packages

Everything below observed/built/tested on `x86_64-linux`, after cherry-picking the commit onto [current `nixos-unstable`](https://github.com/NixOS/nixpkgs/tree/6c9a78c09ff4d6c21d0319114873508a6ec01655).

##### Reviewed points

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on `x86_64-linux`
- [x] executables tested on `x86_64-linux`
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] `meta.mainProgram` is set, if applicable.  (*not applicable here*)
- [x] build time only dependencies are declared in `nativeBuildInputs`  (*I think so...*)
- [x] source is fetched from an official or trusted location
- [x] source is fetched using the appropriate function
- [x] the motives for any special packaging choices are documented  (*none*)
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.  (*none*)
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed  (*none*)
- [x] patches that are remotely available are fetched rather than vendored  (*none*)

More:

- [x] `nixpkgs-review` on `x86_64-linux` (see below)
- [x] CUPS produces printer data without reporting any error, after installing the ppd file with `hardware.ensurePrinters` (tested by printing into a file as I have no Rollo printer)
- [x] `pkgsCross.aarch64-multiplatform.rollo-printer` builds

##### `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`
Commit: `61d7471506d0a76b6553a54edc85bf618f250a7a`

---
###### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>rollo-printer</li>
  </ul>
</details>

shymega commented:
@SuperSandro2000 Seeing as we've had two OKs on this PR, could we add it to the merge queue, please?

Thank you!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2943

r-vdp commented:
CI is falling, we require structured attrs now for new packages.

shymega commented:
Can you elaborate or provide external documentation for the new
structured attributes requirement for new packages?

I can see from CI that its expecting `strictDeps` and structured
attributes, but with no references to documentation that explains *what*
the requirement precisely consists of, which leaves a lot in question.

On 14.04.2026 04:06, Ramses wrote:
>r-vdp left a comment (NixOS/nixpkgs#464258)
>
>CI is falling, we require structured attrs now for new packages.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/NixOS/nixpkgs/pull/464258#issuecomment-4243385118
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

r-vdp commented:
See https://nix.dev/manual/nix/2.34/language/advanced-attributes#adv-attr-structuredAttrs
and https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-strictDeps

Basically you add `__structuredAttrs = true;` and `strictDeps = true;` to the derivation, and check if it still builds, if not, you fix any errors that came up.

Yarny0 reviewed (approved):
I repeated all tests from [my previous review](https://github.com/NixOS/nixpkgs/pull/464258#pullrequestreview-4008492135), after rebasing the [current commit](https://github.com/NixOS/nixpkgs/commit/c2271c2b82c717443d30d1080d55f88757b72869) onto [current `nixos-unstable`](https://github.com/NixOS/nixpkgs/tree/4c1018dae018162ec878d42fec712642d214fdfa).  With the same results.  I reaffirm my approval.

r-vdp committed:
rollo-printer: use pkg-config instead of native cups

configure looks for CUPS via pkg-config first and only falls back to
cups-config. With strictDeps the build-platform cups-config leaks
build-platform -I/-L flags into the cross build (ld warns about
"skipping incompatible libcups.so" and only links because the cc-wrapper
adds the host cups via NIX_LDFLAGS). The cross-aware pkg-config wrapper
picks up the host cups.pc directly, so cups no longer needs to be a
native input.

r-vdp reviewed (approved):
Pushed a small fix for cross builds, LGTM now.
